File: tests/rosetta/transpiler/Pascal/15-puzzle-game.pas
{$mode objfpc}
program Main;
uses SysUtils;
var _nowSeed: int64 = 0;
var _nowSeeded: boolean = false;
procedure init_now();
var s: string; v: int64;
begin
  s := GetEnvironmentVariable('MOCHI_NOW_SEED');
  if s <> '' then begin
    Val(s, v);
    _nowSeed := v;
    _nowSeeded := true;
  end;
end;
function _now(): integer;
begin
  if _nowSeeded then begin
    _nowSeed := (_nowSeed * 1664525 + 1013904223) mod 2147483647;
    _now := _nowSeed;
  end else begin
    _now := Integer(GetTickCount64()*1000);
  end;
end;
function _mem(): int64;
var h: TFPCHeapStatus;
begin
  h := GetFPCHeapStatus;
  _mem := h.CurrHeapUsed;
end;
function _input(): string;
var s: string;
begin
  if EOF(Input) then s := '' else ReadLn(s);
  _input := s;
end;
type MoveResult = record
  idx: integer;
  ok: boolean;
end;
var
  bench_start_0: integer;
  bench_dur_0: integer;
  bench_mem_0: int64;
  bench_memdiff_0: int64;
  board: array of integer;
  solved: array of integer;
  empty: integer;
  moves: integer;
  quit: boolean;
  isSolved_i: integer;
  doMove_r: MoveResult;
  doMove_i: integer;
  doMove_j: integer;
  doMove_tmp: integer;
  shuffle_i: integer;
  printBoard_line: string;
  printBoard_i: integer;
  printBoard_val: integer;
  printBoard_s: string;
  playOneMove_s: string;
  playOneMove_c: string;
  playOneMove_m: integer;
function makeMoveResult(idx: integer; ok: boolean): MoveResult;
begin
  Result.idx := idx;
  Result.ok := ok;
end;
function randMove(): integer;
begin
  exit(_now() mod 4);
end;
function isSolved(): boolean;
begin
  isSolved_i := 0;
  while isSolved_i < 16 do begin
  if board[isSolved_i] <> solved[isSolved_i] then begin
  exit(false);
end;
  isSolved_i := isSolved_i + 1;
end;
  exit(true);
end;
function isValidMove(m: integer): MoveResult;
begin
  if m = 0 then begin
  exit(makeMoveResult(empty - 4, (empty div 4) > 0));
end;
  if m = 1 then begin
  exit(makeMoveResult(empty + 4, (empty div 4) < 3));
end;
  if m = 2 then begin
  exit(makeMoveResult(empty + 1, (empty mod 4) < 3));
end;
  if m = 3 then begin
  exit(makeMoveResult(empty - 1, (empty mod 4) > 0));
end;
  exit(makeMoveResult(0, false));
end;
function doMove(m: integer): boolean;
begin
  doMove_r := isValidMove(m);
  if not doMove_r.ok then begin
  exit(false);
end;
  doMove_i := empty;
  doMove_j := doMove_r.idx;
  doMove_tmp := board[doMove_i];
  board[doMove_i] := board[doMove_j];
  board[doMove_j] := doMove_tmp;
  empty := doMove_j;
  moves := moves + 1;
  exit(true);
end;
procedure shuffle(n: integer);
begin
  shuffle_i := 0;
  while (shuffle_i < n) or isSolved() do begin
  if doMove(randMove()) then begin
  shuffle_i := shuffle_i + 1;
end;
end;
end;
procedure printBoard();
begin
  printBoard_line := '';
  printBoard_i := 0;
  while printBoard_i < 16 do begin
  printBoard_val := board[printBoard_i];
  if printBoard_val = 0 then begin
  printBoard_line := printBoard_line + '  .';
end else begin
  printBoard_s := IntToStr(printBoard_val);
  if printBoard_val < 10 then begin
  printBoard_line := (printBoard_line + '  ') + printBoard_s;
end else begin
  printBoard_line := (printBoard_line + ' ') + printBoard_s;
end;
end;
  if (printBoard_i mod 4) = 3 then begin
  writeln(printBoard_line);
  printBoard_line := '';
end;
  printBoard_i := printBoard_i + 1;
end;
end;
procedure playOneMove();
begin
  while true do begin
  writeln(('Enter move #' + IntToStr(moves + 1)) + ' (U, D, L, R, or Q): ');
  playOneMove_s := _input();
  if playOneMove_s = '' then begin
  continue;
end;
  playOneMove_c := copy(playOneMove_s, 0+1, (1 - 0));
  playOneMove_m := 0;
  if (playOneMove_c = 'U') or (playOneMove_c = 'u') then begin
  playOneMove_m := 0;
end else begin
  if (playOneMove_c = 'D') or (playOneMove_c = 'd') then begin
  playOneMove_m := 1;
end else begin
  if (playOneMove_c = 'R') or (playOneMove_c = 'r') then begin
  playOneMove_m := 2;
end else begin
  if (playOneMove_c = 'L') or (playOneMove_c = 'l') then begin
  playOneMove_m := 3;
end else begin
  if (playOneMove_c = 'Q') or (playOneMove_c = 'q') then begin
  writeln(('Quiting after ' + IntToStr(moves)) + ' moves.');
  quit := true;
  exit();
end else begin
  writeln((('Please enter "U", "D", "L", or "R" to move the empty cell' + #10 + '' + 'up, down, left, or right. You can also enter "Q" to quit.' + #10 + '') + 'Upper or lowercase is accepted and only the first non-blank' + #10 + '') + 'character is important (i.e. you may enter "up" if you like).');
  continue;
end;
end;
end;
end;
end;
  if not doMove(playOneMove_m) then begin
  writeln('That is not a valid move at the moment.');
  continue;
end;
  exit();
end;
end;
procedure play();
begin
  writeln('Starting board:');
  while not quit and (isSolved() = false) do begin
  writeln('');
  printBoard();
  playOneMove();
end;
  if isSolved() then begin
  writeln(('You solved the puzzle in ' + IntToStr(moves)) + ' moves.');
end;
end;
procedure main();
begin
  shuffle(50);
  play();
end;
begin
  init_now();
  bench_mem_0 := _mem();
  bench_start_0 := _now();
  board := [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 0];
  solved := [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 0];
  empty := 15;
  moves := 0;
  quit := false;
  main();
  Sleep(1);
  bench_memdiff_0 := _mem() - bench_mem_0;
  bench_dur_0 := (_now() - bench_start_0) div 1000;
  writeln('{');
  writeln(('  "duration_us": ' + IntToStr(bench_dur_0)) + ',');
  writeln(('  "memory_bytes": ' + IntToStr(bench_memdiff_0)) + ',');
  writeln(('  "name": "' + 'main') + '"');
  writeln('}');
end.

File: tests/rosetta/transpiler/FS/15-puzzle-game.fs
// Generated 2025-07-25 12:29 +0700

exception Break
exception Continue

exception Return

let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
type MoveResult = {
    idx: int
    ok: bool
}
open System

let mutable board: int array = [|1; 2; 3; 4; 5; 6; 7; 8; 9; 10; 11; 12; 13; 14; 15; 0|]
let solved: int array = [|1; 2; 3; 4; 5; 6; 7; 8; 9; 10; 11; 12; 13; 14; 15; 0|]
let mutable empty: int = 15
let mutable moves: int = 0
let mutable quit: bool = false
let rec randMove () =
    let mutable __ret : int = Unchecked.defaultof<int>
    try
        __ret <- (_now()) % 4
        raise Return
        __ret
    with
        | Return -> __ret
and isSolved () =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    try
        let mutable i: int = 0
        while i < 16 do
            if (board.[i]) <> (solved.[i]) then
                __ret <- false
                raise Return
            i <- i + 1
        __ret <- true
        raise Return
        __ret
    with
        | Return -> __ret
and isValidMove (m: int) =
    let mutable __ret : MoveResult = Unchecked.defaultof<MoveResult>
    let mutable m = m
    try
        if m = 0 then
            __ret <- { idx = empty - 4; ok = (empty / 4) > 0 }
            raise Return
        if m = 1 then
            __ret <- { idx = empty + 4; ok = (empty / 4) < 3 }
            raise Return
        if m = 2 then
            __ret <- { idx = empty + 1; ok = (empty % 4) < 3 }
            raise Return
        if m = 3 then
            __ret <- { idx = empty - 1; ok = (empty % 4) > 0 }
            raise Return
        __ret <- { idx = 0; ok = false }
        raise Return
        __ret
    with
        | Return -> __ret
and doMove (m: int) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable m = m
    try
        let r = isValidMove m
        if not (r.ok) then
            __ret <- false
            raise Return
        let i: int = empty
        let j = r.idx
        let tmp = board.[i]
        board.[i] <- board.[j]
        board.[j] <- tmp
        empty <- j
        moves <- moves + 1
        __ret <- true
        raise Return
        __ret
    with
        | Return -> __ret
and shuffle (n: int) =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    let mutable n = n
    try
        let mutable i: int = 0
        while (i < n) || (isSolved()) do
            if doMove (randMove()) then
                i <- i + 1
        __ret
    with
        | Return -> __ret
and printBoard () =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    try
        let mutable line: string = ""
        let mutable i: int = 0
        while i < 16 do
            let ``val`` = board.[i]
            if ``val`` = 0 then
                line <- line + "  ."
            else
                let s: string = string ``val``
                if ``val`` < 10 then
                    line <- (line + "  ") + s
                else
                    line <- (line + " ") + s
            if (i % 4) = 3 then
                printfn "%s" line
                line <- ""
            i <- i + 1
        __ret
    with
        | Return -> __ret
and playOneMove () =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    try
        try
            while true do
                printfn "%s" (("Enter move #" + (string (moves + 1))) + " (U, D, L, R, or Q): ")
                let s: string = System.Console.ReadLine()
                if s = "" then
                    raise Continue
                let c: string = s.Substring(0, 1 - 0)
                let mutable m: int = 0
                if (c = "U") || (c = "u") then
                    m <- 0
                else
                    if (c = "D") || (c = "d") then
                        m <- 1
                    else
                        if (c = "R") || (c = "r") then
                            m <- 2
                        else
                            if (c = "L") || (c = "l") then
                                m <- 3
                            else
                                if (c = "Q") || (c = "q") then
                                    printfn "%s" (("Quiting after " + (string moves)) + " moves.")
                                    quit <- true
                                    __ret <- ()
                                    raise Return
                                else
                                    printfn "%s" ((("Please enter \"U\", \"D\", \"L\", or \"R\" to move the empty cell\n" + "up, down, left, or right. You can also enter \"Q\" to quit.\n") + "Upper or lowercase is accepted and only the first non-blank\n") + "character is important (i.e. you may enter \"up\" if you like).")
                                    raise Continue
                if not (doMove m) then
                    printfn "%s" "That is not a valid move at the moment."
                    raise Continue
                __ret <- ()
                raise Return
        with
        | Break -> ()
        | Continue -> ()
        __ret
    with
        | Return -> __ret
and play () =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    try
        printfn "%s" "Starting board:"
        while (not quit) && ((isSolved()) = false) do
            printfn "%s" ""
            printBoard()
            playOneMove()
        if isSolved() then
            printfn "%s" (("You solved the puzzle in " + (string moves)) + " moves.")
        __ret
    with
        | Return -> __ret
and main () =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    try
        let __bench_start = _now()
        let __mem_start = System.GC.GetTotalMemory(true)
        shuffle 50
        play()
        let __bench_end = _now()
        let __mem_end = System.GC.GetTotalMemory(true)
        printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

        __ret
    with
        | Return -> __ret
main()
